### PR TITLE
Feat/gen table

### PR DESCRIPTION
**Pull request by dmoore247**
Remove properties generated by TSQL that HIve/Spark/Databricks don't support.
- clustered Primary keys
- clustered storage 
etc.

### CODE REVIEW

File: sqlglot/dialects/hive.py [hunk 1]
```diff
@@ -466,6 +477,12 @@ class Generator(generator.Generator):
             exp.NumberToStr: rename_func("FORMAT_NUMBER"),
             exp.LastDateOfMonth: rename_func("LAST_DAY"),
             exp.National: lambda self, e: self.national_sql(e, prefix=""),
+            exp.Constraint: constraint_sql,
+            exp.ClusteredColumnConstraint: lambda self, e: f"({self.expressions(e, 'this', indent=False)})",
```
Comment (tobymao) [inline sqlglot/dialects/hive.py:481]: what are you trying to do here? this looks wrong

File: sqlglot/dialects/hive.py [hunk 2]
```diff
@@ -51,6 +51,17 @@
 DIFF_MONTH_SWITCH = ("YEAR", "QUARTER", "MONTH")
 
 
+def constraint_sql(self, expression: exp.Constraint) -> str:
+    this = self.sql(expression, "this")
+
+    # remove unsupported properties
+    expression.expressions.remove(expression.find(exp.Properties))
```
Comment (tobymao) [inline sqlglot/dialects/hive.py:58]: i don't understand what this is doing, it doesn't look correct

File: tests/dialects/test_tsql.py [hunk 3]
```diff
@@ -18,16 +18,28 @@ def test_tsql(self):
...
-            ) ON [PRIMARY]
+                [zip_cd] [varchar](5) NULL NOT FOR REPLICATION,
+                [zip_cd_mkey] [varchar](5) NOT NULL,
+                CONSTRAINT [pk_mytable] PRIMARY KEY CLUSTERED ([zip_cd_mkey] ASC)
+                WITH (PAD_INDEX = ON, STATISTICS_NORECOMPUTE = OFF) ON [INDEX]
+            ) ON [SECONDARY]
             """,
-            'CREATE TABLE x ("zip_cd" VARCHAR(5) NULL NOT FOR REPLICATION CONSTRAINT "pk_mytable" PRIMARY KEY CLUSTERED ("zip_cd_mkey") WITH (PAD_INDEX=ON, STATISTICS_NORECOMPUTE=OFF) ON "PRIMARY") ON "PRIMARY"',
+            write={
+                #'tsql':'CREATE TABLE x ("zip_cd" VARCHAR(5) NULL NOT FOR REPLICATION CONSTRAINT "pk_mytable" PRIMARY KEY CLUSTERED ("zip_cd_mkey") WITH (PAD_INDEX=ON, STATISTICS_NORECOMPUTE=OFF) ON "PRIMARY") ON "PRIMARY"',
```
Comment (tobymao) [inline tests/dialects/test_tsql.py:31]: uncomment this

File: sqlglot/dialects/hive.py (hunk 2 above)
Comment (dmoore247) [reply sqlglot/dialects/hive.py:58]: Hive / Spark support some constraints but not all constraints other dialects that might emit.

This code intends to filter out the non-supported constraints, so this code does a find and removes them.

Comment (tobymao) [reply sqlglot/dialects/hive.py:58]: what constraints are being removed though? just all properties?, i think you can just do

for prop in expression.find_all(exp.Properties):
    prop.pop()

**Comment by dmoore247** [comment]
We're waiting for your response to question above...

Is remove bad? It worked better than iterating over a collection and popping from the same collection which appeared to create a race condition.
Is the concern that properties are removed or how the properties are being removed?

**Comment by tobymao** [comment]
what question?

File: sqlglot/dialects/hive.py (hunk 2 above)
Comment (dmoore247) [reply sqlglot/dialects/hive.py:58]: Is `remove` bad? It worked better than iterating over a collection and popping from the same collection which appeared to create a race condition.
Is the concern that properties are removed or how the properties are being removed?

**Review by dmoore247 — COMMENTED** [review]
Review complete

File: sqlglot/dialects/hive.py (hunk 2 above)
Comment (tobymao) [reply sqlglot/dialects/hive.py:58]: this is not resolved, you didn't answer the question about why this is being removed.

Comment (tobymao) [reply sqlglot/dialects/hive.py:58]: it seems too broad, and it should use pop, and not remove

File: sqlglot/dialects/hive.py [hunk 4]
```diff
@@ -51,6 +51,17 @@
 DIFF_MONTH_SWITCH = ("YEAR", "QUARTER", "MONTH")
 
 
+def constraint_sql(self, expression: exp.Constraint) -> str:
+    this = self.sql(expression, "this")
```
Comment (tobymao) [inline sqlglot/dialects/hive.py:55]: you need to call copy up here, move this function into the generator directly and clean it up

File: sqlglot/dialects/hive.py [hunk 5]
```diff
@@ -466,6 +477,12 @@ class Generator(generator.Generator):
             exp.NumberToStr: rename_func("FORMAT_NUMBER"),
             exp.LastDateOfMonth: rename_func("LAST_DAY"),
             exp.National: lambda self, e: self.national_sql(e, prefix=""),
+            exp.Constraint: constraint_sql,
+            exp.ClusteredColumnConstraint: lambda self, e: f"({self.expressions(e, 'this', indent=False)})",
+            exp.NonClusteredColumnConstraint: lambda self, e: f"({self.expressions(e, 'this', indent=False)})",
+            exp.NotForReplicationColumnConstraint: lambda self, e: "",
```
Comment (tobymao) [inline sqlglot/dialects/hive.py:483]: i think you could just change the properties location intsead of doing this

Comment (tobymao) [reply sqlglot/dialects/hive.py:483]: change it to UNSUPPORTED location property

Comment (dmoore247) [reply sqlglot/dialects/hive.py:483]: I'm really not understanding this feedback.

Do you mean the "property's location" or the "location property"?
'location' property as in table location as in the storage path, and how does that relate to a clustered constraint?

Comment (tobymao) [reply sqlglot/dialects/hive.py:483]: look up Properties.Location and how it's used, switch it to UNSUPPORTED

Comment (dmoore247) [reply sqlglot/dialects/hive.py:483]: The pattern appears to be to return "" for unsupported options.

File: sqlglot/dialects/hive.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmoore247) [reply sqlglot/dialects/hive.py:481]: Let me research

File: sqlglot/dialects/hive.py (hunk 4 above)
Comment (dmoore247) [reply sqlglot/dialects/hive.py:55]: Copy the expression object? and 
in-line the code with the Generator list for more concise code?

Comment (tobymao) [reply sqlglot/dialects/hive.py:55]: yea there’s no need to pass this in

File: sqlglot/dialects/hive.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmoore247) [reply sqlglot/dialects/hive.py:481]: Essentially removing the 'CLUSTERED' keyword from the generation.

File: sqlglot/dialects/hive.py (hunk 4 above)
Comment (dmoore247) [reply sqlglot/dialects/hive.py:55]: The `this` is the name of the constraint. Removing this removes the constraint name.

```python
def constraint_sql(self, expression: exp.Constraint) -> str:
    expression = expression.copy()
    this = self.sql(expression, "this")

    # remove unsupported properties
    expression.expressions.remove(expression.find(exp.Properties))
    expression.expressions.remove(expression.find(exp.OnProperty))

    expressions = self.expressions(expression, sep=" ", flat=True)
    return f"CONSTRAINT {this} {expressions}"
    
```

**Comment by dmoore247** [comment]
Toby, is there anything else you need?

File: tests/dialects/test_tsql.py [hunk 6]
```diff
@@ -18,16 +18,29 @@ def test_tsql(self):
             'CREATE TABLE x (CONSTRAINT "pk_mytable" UNIQUE NONCLUSTERED (a DESC)) ON b (c)'
         )
 
-        self.validate_identity(
+        self.maxDiff = None
```
Comment (tobymao) [inline tests/dialects/test_tsql.py:21]: delete this maxDiff

Comment (dmoore247) [reply tests/dialects/test_tsql.py]: Done

**Pull request merged**